Provide framework for implementing VACs

albireox opened the pull request:
I've made some modifications to the `vacs` branch. Still work in progress but this PR:

- [X] Restructures the VAC framework to simplify the code the VAC owners need to write.
- [X] Provides two methods to download the VAC files: either use an entry in `tree/sdss_paths.ini` or use a relative path in `mangawork/sandbox`.
- [X] Provides a framework for VACs that are composed of many files (e.g., galaxyzoo3d).
- [X] A check is performed when the VAC is being accessed. If the file does not exists, it is downloaded automatically.
- [X] Provides an example of how to use sdss_access (dapall) an another of how to use a custom-defined sandbox path (galaxyzoo3d).
- [X] All VACs are grouped under an attribute, e.g., `my_maps.vacc.galaxyzoo3d`.
- [X] `VACMixIn` is now inserted in `Cube`, `Maps`, and `ModelCube` since it does not make sense for all classes that inherit from `MarvinToolsClass`.
- [x] Provides tests.
- [X] Provides documentation for VAC owners.
- [x] Provides documentation on how to access VACs from the tools.
- [x] Provide an (auto-generated?) list of available VACs.
- [x] `dapall` should be removed from VACs. Leave file as an example?
- [x] Adds VACs into the DataModel (by release) as a lookup. 
- [x] Basic VAC DataModel included in Sphinx Documentation
- [ ] Request review of the Galaxy Zoo: 3D VAC from Karen and Coleman.


Relevant documentation is [here](https://sdss-marvin.readthedocs.io/en/vacs/contributing/contributing.html#) and [here](https://sdss-marvin.readthedocs.io/en/vacs/tools/metadata.html).

Closes #433 
Closes #240

havok2063 commented on docs/sphinx/contributing/contributing.rst:
```diff
@@ -0,0 +1,150 @@
+* If you have used Marvin in the classroom and developed activities based on it we would love to hear about it!
+
+
+Contributing a VAC
```
We should make sure this fits in with the real `Contributing a VAC` guidelines the Data Team have and make it clear this is optional for people with MaNGA VACS.  It should also include a timeline placed on the VAC owners, similar to the DR timeline, that places these steps in context relative to those.    

This documentation implies a lot of steps needed for VAC inclusion.  I think that's fine and necessary, but it may be unreasonable now to expect VAC owners to deliver this content on a timeline for inclusion in DR15.  Especially given writing Python-code, tests, Sphinx documentation, will be all new to most of these people.  And given the Data Team's usual problems with corralling VAC owners to provide what the "simple things" they need for the DR release.  This will require extra time for us to vet everything, which I doubt we have the time and/or manpower for.

albireox replied:
I honestly don't think we need to require tests or documentation for the VACs. Maybe that's not clear in the documentation. Honestly, as long as we make clear the VACs are not officially supported by us or MaNGA, and they don't break anything else, I'd leave it to the VAC owners. Chances of us getting more than a couple VACs otherwise are very remote.

Also, not sure we need to tie this to any DR timeline. We are not even including VACs in the database and there is no server counterpart for this. A new VAC can be implemented in any new release of Marvin.

havok2063 replied:
VACs are officially supported.  Not officially maintained but MaNGA VACs have to get official approval and vetting by the DPC before they are allowed to be released.  And they have to adhere to the official VAC guidelines by the Data Team before they are allowed to be released.  Documentation is one of the mandatory guidelines.  We should be the same.  

Maybe we don't need to require tests but that's the only real way to ensure they don't break anything else.  If we can write some simple example tests on the `helloworld` example they can emulate, that might be enough.  

VACs are only released during Data Releases and not in-between.  Even internally.  I think we should adopt a similar policy.  New VACs will only be included during a release of Marvin for the Data Release when the VACs themselves are being released.  It keeps things a bit cleaner because that's when VAC owners are preparing their VACs.  Even if we roll out VACs in releases after a DR, I think it's better to impose a relative timeline.  Otherwise, people will never do it or take forever.  I think Anne-Marie would agree with me here that people need deadlines placed on them.

havok2063 commented on python/marvin/contrib/vacs/base.py:
```diff
@@ -0,0 +1,172 @@
+
+        self.rsync_access.templates[self.name] = os.path.join('$MANGA_SANDBOX', relative_path)
+
+    def get_path(self, path_params={}):
```
Additionally I think we don't need this as all officially released VACS live on the SAS in one official place, the SANDBOX which sdss_access already knows about it.  Marvin will only support officially approved and released VACs.

albireox replied:
Yeah, this is one of the things we should talk about. In general I agree, but I'm fairly concerned that having to create a PR in tree for a new path, understand the functionality there, etc may increase the steepness of contributing a VAC. I'm also not 100% sure we want to overcrowd the tree with VACs, many of which have filename format that are not 100% easy to implement with the current framework. What I was trying is to provide a simple way to use sdss_access to download a given path from the SAS without having to permanently add it to the tree. Anyway, let's discuss this on Monday.

havok2063 replied:
Since VACS only come around during DRs, and only in handfuls, I think we can get away with having one branch of tree called vacs and have people commit into that.  They are only adding one line in one line.  Given everything else they have to do for their VAC in general, I think that's ok.  

sdss_access isn't just for official sdss data products.  It's for every single file housed on the SAS; every file that has a datamodel html page.  If you look through the ini file, you'll find tons of intermediate and non-official files.  Plus, VACS have to create a datamodel html file as part of their release, and they become permanently housed by the SAS upon release, so I think it's fine to add this step in there for them.  If they have unconventional naming formats, I think that's even a stronger case to put it in sdss_access.

albireox replied:
That's a fair point. I want to make implementation of VACs in Marvin as easy as possible but make part of the VAC owner responsibilities before release to add a path in `sdss_paths.ini` seems like a good idea.

albireox commented on python/marvin/contrib/vacs/dapall.py:
```diff
@@ -0,0 +1,67 @@
+    """
+
+    name = 'dapall'
+    path_params = {'dapver': '2.1.3', 'drpver': 'v2_3_1'}
```
For the DAP that's easy because the versions are those that we use in Marvin anyway (the DAP is really a bad example, maybe we should not use it at all). But how are we supposed to do that for, for instance, the version of the Galaxy Zoo catalogue?

havok2063 replied:
I was also thinking about removing it since we have already included it, and it might be confusing for people.  But I'm flexible.  In my original idea, the user would have to download the vacs manually first.  They would then manually specify the parameters to the filepath.  Once the vac files were local, then they automatically get swept up into the container by Marvin.  But if you want the downloading to happen automatically we can discuss this more.

albireox replied:
If we require the VACs to have to be in the tree then I think there is little difference between the DAPall and galaxyzoo3d, so I think we can just remove the former and make it clearer.

I'm quite opposed to having to download the VAC manually. I think it defeats the purpose of easy access to VACs. I think we should link a release (MPL-6, DR15) to a VAC version but not go beyond that.

havok2063 replied:
I like linking VAC versions to a release.  That makes sense to me, especially since they are released for a specific release.  Then we can tell exactly what vacs became available when, plus we have some versioning which we can use to automatically download them.  We should build this into the DataModel for a specific release, so we know which VACs (and versions) are available in which release.

albireox commented on python/marvin/tools/maps.py:
```diff
@@ -116,6 +114,8 @@ def __init__(self, input=None, filename=None, mangaid=None, plateifu=None,
 
         self._check_versions(self)
 
+        self.vacs = VACMixIn.get_vacs(self)
```
I think it should not be added to Spaxel and Bin for the same reason we don't add the NSA or dapall information there. Once you get a bin or spaxel I think we should remove galaxy-wide metadata.

I'm almost ok with adding it directly to `MarvinToolsClass`, the problem being that `Plate` also inherits from it. Maybe we can manually remove it in `Plate` but it seems a bit like a hack.

havok2063 replied:
I only bring this up because Kate explicitly asked in our session if VACS would be available on the Spaxels.  I know her VAC consists of individual spaxel classifications and measurements.   She'll have two files, one of which consists of rows of spaxels not galaxies.    

Yeah, when I made it didn't make much sense for it to be subclassed from `MarvinToolsClass` but I wanted to take advantage of the MMA that it had.  But maybe we can refactor `Plate`.  I've never been entirely happy with it.  But I don't see a huge problem with having vacs attached to `Plate`.  There are only a handful of basic parameters users will need to select rows in their vacs.  We can probably build something into the `Base` that checks for these depending on which `Tool` is being passed.

albireox replied:
I'd like to see an example of her files and what she'd return before creating a framework for that, but I don't think it's undoable.

I did try to implement a system in which the subclass of `VACMixIn` was able to define in which Marvin classes it wanted to show up but I run into 1001 circular import errors. I may give it another try.

havok2063 replied:
I can talk to her about her files again and see if she has produced example versions.  I know she has two kinds of files.  One is a single fits file where each row is an individual spaxel in a galaxy and contains columns like `plateifu`, `bintype`, `template`, `x`, `y` and all of her spaxel classifications.  So this file has millions of rows, spanning all spaxels for all galaxies.   So my guess is each `Spaxel` would return that vac info (PCA classes and params) for that matching row.   

Her other filetype is a directory containing one fits file per galaxy, and it contains basically a new map of her PCA classifications.  So this VAC info would be returned by a `Cube` or `Map`

albireox replied:
I have added a global `include` attribute that defines in which tool classes the VAC should show up. It always skips the `Plate`. I think that should provide the framework for Kate's VAC. `VACMixIn` now is loaded in the `MarvinToolsClass` and in `SpaxelBase`.

albireox commented on python/marvin/contrib/vacs/__init__.py:
```diff
@@ -0,0 +1,23 @@
+import os
+import pkgutil
+
+from marvin.contrib.vacs.base import VACMixIn  # noqa
```
I think it does, I'll try again with the metaclass.

albireox commented on python/marvin/contrib/vacs/dapall.py:
```diff
@@ -0,0 +1,67 @@
+    name = 'dapall'
+    path_params = {'dapver': '2.1.3', 'drpver': 'v2_3_1'}
+
+    def get_data(self, parent_object):
```
I think the hello world example should be a real one. The galaxyzoo3d one is actually fairly trivial. Have a look at the [documentation](https://sdss-marvin.readthedocs.io/en/vacs/contributing/contributing.html#contributing-a-vac) and let me know if that makes it clearer.

albireox commented on python/marvin/contrib/vacs/dapall.py:
```diff
@@ -0,0 +1,67 @@
+
+        dap_table = astropy.table.Table.read(filename)
+
+        plate = parent_object.plate
```
Is there a plateifu? Yeah, then that would be best. I'll change that.

albireox commented on python/marvin/contrib/vacs/galaxyzoo3d.py:
```diff
@@ -0,0 +1,47 @@
+    name = 'galaxyzoo3d'
+    version = 'v1_0_0'
+
+    def get_data(self, parent_object):
```
Actually I cannot find an entry for galaxyzoo3d in sdss_paths.ini. How is it called?

havok2063 replied:
I made a separate branch of tree called `galaxyzoo3d` where we added the name.  Here is the entry we added in that branch, `mangagalaxyzoo3d = $MANGA_SANDBOX/galaxyzoo3d/{gz3dver}/{mangaid}_{ifusize}_{zooid}.fits.gz`

albireox replied:
Is there a definition of how `zooid` is calculated in `sdss_access`? Otherwise this definition becomes harder to use than a self-generated path.

havok2063 replied:
`zooid` isn't calculated.  If I recall, that is a random number Coleman made up.  Or it may have some internal significance but that's it.  In his code, it's an input.  Users would have to look this number up whether they generated the path themselves or used sdss_access.  We might not need the `zooid`.  The `mangaid` and `ifusize` is enough to pick out a unique file.  Setting `zooid='*'` would probably work.  We should double-check with Coleman though.

albireox replied:
Great. Random numbers in paths. Something else? A sphinx asking you three questions before you can download. Hurdles?

albireox commented on python/marvin/core/core.py:
```diff
@@ -60,7 +60,7 @@ def kwargsGet(kwargs, key, replacement):
 
 
-class MarvinToolsClass(object, six.with_metaclass(abc.ABCMeta)):
+class MarvinToolsClass(object):
```
Ok, I reenabled that. Not sure exactly what I was trying to do when it failed but seems to work now.

albireox commented:
I'm made some changes based on Brian's input:

- I've removed the `set_sandbox_path` method and now everything must be downloaded after adding the path to `sdss_paths.ini`. I'm afraid that is going to become a significant bottleneck but maybe I'm not giving people enough credit.
- I added test to make sure that all the VACs return "something" and they don't break Marvin. My opinion is that must be our main concern and checking that the VAC actually returns the expected data must fall on the VAC owners.
- I removed the DAPall completely from the VACs and the documentation. The only example now refers to Galaxy Zoo 3D.
- There is an `include` attribute that defines in which classes the VAC will be included. 
- The galaxyzoo3d example shows hot to link the Marvin release with a VAC internal version.

havok2063 commented:
After thinking about this a bit more, I'd like to keep the number of pieces of code users have to modify to get a VAC integrated down to a small number.  So I've kept the version for each VAC included in the VAC.  I think that's a nice self-contained place for it.  So that VACs can be looked up in a given release, I've added a simple VAC datamodel that bundles all available vacs by release.  The VAC datamodel is also auto-documented into the DataModel section of the Sphinx docs under each release.   For now, it shows the name, version, and description.  Later, if we want we can be a bit fancy here so that a vac datamodel provides more detailed info.

albireox commented:
This looks good to me. I like the idea of having everything self contained but still having the ability of using a datamodel.

I removed some old gz3d utilities that I think had been there since December. Those can be reimplemented as a contributed tool but I don't think they should be a core part of Marvin.

I think the only thing remaining before merging this is deciding what to do about the galaxy zoo VAC. We have implemented it and I used it to write the documentation but we have decided to not include VACs in DR15. Maybe we should wait and not merge this until after DR15.

havok2063 commented:
I think we're nearly finished here.  I left in Coleman's gz3d code because that was his code that actually provides the useful information about his vac to the user.  The actual VAC files that we return in the example aren't the desired endpoint for the user if I recall from talking to him.  That's why this example may be a bit more complicated.  He may not want to return a simple row or Table as we do but pass back this object that generates different masks.   We need to have another discussion with him and Karen.  There's also the fact that this VAC will not be officially released in DR15.  We may want to consider building an example out of the HI VAC from Dave Stark and Karen instead.

havok2063 commented:
I'm going to merge this branch and merge into the `vetround2` branch.
